File: alternative-reality/alt.c

#include <stdio.h>

int coins[] = { 1, 5, 10, 25, 50 };
int cache[512];

static int
count_coins(int target, int num_coins)
{
  int cache_index, result;

  if (target < 0 || num_coins <= 0) {
    return 0;
  }
  if (target == 0) {
    return 1;
  }

  cache_index = target*5 + num_coins;
  if (cache[cache_index]) {
    return cache[cache_index];
  }

  result = count_coins(target, num_coins-1)
         + count_coins(target - coins[num_coins-1], num_coins);
  cache[cache_index] = result;

  return result;
}

int
main(int argc, char const *argv[])
{
  FILE *file = fopen(argv[1], "r");
  int target;
  while (fscanf(file, " %d ", &target) > 0) {
    printf("%d\n", count_coins(target, 5));
  }
  return 0;
}

File: alternative-reality/alt.rb
module AlternativeReality
  COINS = [ 1, 5, 10, 25, 50 ]
  CACHE = Array.new(512)

  def self.count_coins(target, num_coins)
    return 0 if target < 0 || num_coins <= 0
    return 1 if target == 0

    cache_index = target*5 + num_coins
    return CACHE[cache_index] if CACHE[cache_index]

    result = count_coins(target, num_coins-1) +
             count_coins(target - COINS[num_coins-1], num_coins)
    CACHE[cache_index] = result

    result
  end

  def self.run(filename = ARGV[0])
    File.open(filename) do |file|
      file.each_line do |line|
        line.strip!
        next if line.empty?

        puts count_coins(line.to_i, 5)
      end
    end
  end
end

AlternativeReality.run if __FILE__ == $0